utils/raspberrypi/ctt/ctt_ccm.py

```diff
@@ -87,48 +117,128 @@ def ccm(Cam, cal_cr_list, cal_cb_list):
     """
     ccm_tab = {}
     for Img in imgs:
-        Cam.log += '\nProcessing image: ' + Img.name
+        print('\nProcessing image: ' + Img.name)
         """
         get macbeth patches with alsc applied if alsc enabled.
         Note: if alsc is disabled then colour_cals will be set to None and no
         the function will simply return the macbeth patches
         """
-        r, b, g = get_alsc_patches(Img, colour_cals, grey=False)
+        r, b, g = get_alsc_patches(Img, colour_cals, grey=False)  # 256 values for each patch of sRGB values
+
         """
         do awb
         Note: awb is done by measuring the macbeth chart in the image, rather
         than from the awb calibration. This is done so the awb will be perfect
         and the ccm matrices will be more accurate.
         """
         r_greys, b_greys, g_greys = r[3::4], b[3::4], g[3::4]
-        r_g = np.mean(r_greys/g_greys)
-        b_g = np.mean(b_greys/g_greys)
+        r_g = np.mean(r_greys / g_greys)
+        b_g = np.mean(b_greys / g_greys)
         r = r / r_g
         b = b / b_g
-
         """
         normalise brightness wrt reference macbeth colours and then average
         each channel for each patch
         """
-        gain = np.mean(m_srgb)/np.mean((r, g, b))
+        gain = np.mean(m_srgb) / np.mean((r, g, b))
         Cam.log += '\nGain with respect to standard colours: {:.3f}'.format(gain)
-        r = np.mean(gain*r, axis=1)
-        b = np.mean(gain*b, axis=1)
-        g = np.mean(gain*g, axis=1)
-
+        r = np.mean(gain * r, axis=1)
+        b = np.mean(gain * b, axis=1)
+        g = np.mean(gain * g, axis=1)
         """
         calculate ccm matrix
         """
-        ccm = do_ccm(r, g, b, m_srgb)
+        # ==== All of below should in sRGB ===##
+        sumde = 0
+        ccm = do_ccm(r, g, b, m_srgb)  # This is the initial guess that our optimisation code works with.
+
+        r1 = ccm[0]
+        r2 = ccm[1]
+        g1 = ccm[3]
+        g2 = ccm[4]
+        b1 = ccm[6]
+        b2 = ccm[7]
+        '''
+        COLOR MATRIX LOOKS AS BELOW
+        R1 R2 R3   Rval     Outr
+        G1 G2 G3  *  Gval  =  G
+        B1 B2 B3   Bval     B
+        Will be optimising 6 elements and working out the third element using 1-r1-r2 = r3
+        '''
+
+        x0 = [r1, r2, g1, g2, b1, b2]
+        x0 = [2, -1, -0.4, 1.7, 0, -0.9]
+        result = minimize(guess, x0, args=(r, g, b, m_lab), tol=0.0000000001)  # This produces a color matrix which has the lowest delta E possible, based off the input data. Note it is impossible for this to reach zero since inperfect data.
+
+        Cam.log += ("\n \n Optimised Matrix Below: \n \n")
+        [r1, r2, g1, g2, b1, b2] = result.x
+        occm = [r1, r2, (1 - r1 - r2), g1, g2, (1 - g1 - g2), b1, b2, (1 - b1 - b2)]  # This is the optimised Color Matrix
+        Cam.log += str(occm)
+        Cam.log += str(ccm)
```

Comment:
Would the log be easier to follow if it just added that this is the original colour matrix?